File: lang/nodejs/src/fun/vac/data_structures/LinkedStack.js
"use strict"

let Node = class {

    constructor(element) {
        this.data = element
        this.next = null
    }
}

exports.LinkedStack = class {

    constructor(physicalSize = 0) {
        this.head = null
        this.tail = null
        this.length = 0
    }

    size() {
        return this.length
    }

    isEmpty() {
        return this.length === 0
    }

    peek() {
        if (this.length === 0) {
            throw new Error("[PANIC - NoSuchElement]")
        }

        return this.tail.data
    }

    push(element) {
        let node = new Node(element)

        if (this.length === 0) {
            this.head = node
        } else {
            this.tail.next = node
        }

        this.tail = node

        this.length++
    }

    pop() {
        if (this.length === 0) {
            throw new Error("[PANIC - NoSuchElement]")
        }

        let target = this.tail

        if (this.length === 1) {
            this.head = null
            this.tail = null
        } else {
            let cursor = this.head
            for (let i = 0, bound = this.length - 2; i < bound; i++) {
                cursor = cursor.next
            }
            cursor.next = null
            this.tail = cursor
        }

        target.data = null

        this.length--
    }
}

File: lang/python/src/fun/vac/data_structures/LinkedStack.py
class Node:
    def __init__(self, element: object):
        self.data = element
        self.next = None

class LinkedStack:
    def __init__(self):
        self.head = None
        self.tail = None
        self.length = 0

    def size(self) -> int:
        return self.length

    def isEmpty(self) -> bool:
        return self.length == 0

    def peek(self) -> int:
        if self.length == 0:
            raise RuntimeError("[PANIC - NoSuchElement]")

        return self.tail.data

    def push(self, element: object) -> None:
        node = Node(element)

        if self.length == 0:
            self.head = node
        else:
            self.tail.next = node

        self.tail = node

        self.length += 1

    def pop(self) -> None:
        if self.length == 0:
            raise RuntimeError("[PANIC - NoSuchElement]")

        target = self.tail

        if self.length == 1:
            self.head = None
            self.tail = None
        else:
            cursor = self.head
            for i in range(self.length - 2):
                cursor = cursor.next
            cursor.next = None
            self.tail = cursor

        target.data = None

        self.length -= 1
